scipy/sparse/linalg/_matfuncs.py

```diff
@@ -861,3 +862,47 @@ def _ell(A, m):
     log2_alpha_div_u = np.log2(alpha/u)
     value = int(np.ceil(log2_alpha_div_u / (2 * m)))
     return max(value, 0)
+
+def matrix_power(A, power, structure=None):
+    """
+    Raise a square matrix to the integer power, `power`.
+
+    For non-negative integers, ``A**power`` is computed using repeated
+    matrix multiplications. Negative integers are not supported. 
+
+    Parameters
+    ----------
+    A : (M, M) square sparse array or matrix
+        sparse array that will be raised to power `power`
+    power : int
+        Exponent used to raise sparse array `A`
+
+    Returns
+    -------
+    A**power : (M, M) sparse array or matrix
+        The output matrix will be the same shape as A, and will preserve
+        the class of A, but the format of the output may be changed.
+
+    """
+    M, N = A.shape
+    if M != N:
+        raise TypeError('sparse matrix is not square')
+
+    if isintlike(power):
+        power = int(power)
+        if power < 0:
+            raise ValueError('exponent must be >= 0')
+
+        if power == 0:
+            return eye(M, dtype=A.dtype)
+
+        if power == 1:
+            return A.copy()
+
+        tmp = matrix_power(A, power // 2)
+        if power % 2:
+            return A @ tmp @ tmp
+        else:
+            return tmp @ tmp
```

Comment:
What is the benefit of proceeding as follow, here?

Reply:
This is part of the [existing matrix_power implementation](https://github.com/scipy/scipy/pull/513), which (as I understand it) was used to reduce the number of required multiplications. This performance improvement was significant then, and [seems so again.](https://github.com/scipy/scipy/pull/18544#issuecomment-1572118469)

Reply:
Ah yes, that's just a recursive implementation.

We could further reduce the cost of this function using inline computations with more or less `np.log2(power)` iterations not to have the checks just above be rerun several times.

What do you think?